sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+          seek_to_start=seek_to_start).name
+
+    reader = PatchedPubSubReader(self, self._reader_class, **reader_kwargs)
+    return reader
+
+  def writer(self):
+    self._timestamp = time.time()
+    writer = PatchedPubSubWriter(self, self._writer_class, (self.location,),
+                                 self._writer_kwargs)
+    return writer
+
+  @contextlib.contextmanager
+  def read_to_queue(self,
+                    seek_to_start=True,
+                    **kwargs):
+    self._assert_topic_exists()
+
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+
+    if "subscription" in reader_kwargs:
+      created_subsciption = None
+    else:
+      created_subsciption = self._create_child_subscription(
+          seek_to_start=seek_to_start)
+      reader_kwargs["subscription"] = created_subsciption.name
+      assert created_subsciption in self._child_subscriptions
+
+    @functools.total_ordering
+    class PrioritizedTimestampedValue(TimestampedValue):
+
+      def __lt__(self, other):
+        return self.timestamp < other.timestamp
+
+    # Set arbitrary queue size limit to prevent OOM errors.
+    parsed_message_queue = queue.PriorityQueue(1000)
+    decoder = DecodeFromPubSub(
+        self.coder,
+        with_attributes=reader_kwargs.get("with_attributes", False),
+        timestamp_attribute=reader_kwargs["timestamp_attribute"])
+
+    def callback(msg):
+      msg.ack()
+      message = PubsubMessage._from_message(msg)
+      timestamped_value = next(decoder.process(message))
+      ordered_timestamped_value = PrioritizedTimestampedValue(
+          timestamped_value.value, timestamped_value.timestamp)
+      parsed_message_queue.put(ordered_timestamped_value)
+
+    sub_client = pubsub.SubscriberClient()
+    future = sub_client.subscribe(
+        reader_kwargs["subscription"], callback=callback)
+    try:
+      yield parsed_message_queue
+    finally:
+      future.cancel()
+      if created_subsciption is not None:
+        sub_client.delete_subscription(created_subsciption.name)
+        self._child_subscriptions.remove(created_subsciption)
+
+  def read(self, delay=0, timeout=5, **kwargs):
+    with self.read_to_queue(**kwargs) as message_queue:
+      time.sleep(delay)
+      while True:
+        try:
+          element = message_queue.get(timeout=timeout)
+          yield element
+        except queue.Empty:
+          return
+
+  def write(self, elements):
+    if self.requires_coder and self.coder is None:
+      # TODO(ostrokach): We might want to infer the element type from the first
+      # N elements, rather than reading the entire iterator.
+      elements = list(elements)
+      element_type = datatype_inference.infer_element_type(elements)
+      self.coder = coders.registry.get_coder(element_type)
+
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
```

Comment:
What is this keyword argument? I couldn't find documentation for it.

Reply:
I added a docstring which includes the description of this argument. Please let me know if you think it is not necessary and should be removed.